### ISSUE TITLE
Bug in loading keras model with custom losses (metrics)

### ISSUE THREAD

**Issue by mantonios107** [tensorflow/tensorflow#44210]
**System information**
- Have I written custom code: yes, class for custom loss `tf.keras` 
- OS Platform and Distribution Linux Ubuntu 20.04:
- TensorFlow installed from binary
- TensorFlow version 2.1.0 --> 2.3.0 (it seems that the bug is present until master branch):
- Python version: 3.7.6

**Describe the current behavior**

(see code below)
1. Define a class for custom loss in keras named `BatchMeanSquaredError`
2. Create a model using `tf.keras.Sequential`
3. Train the model on random data
4. Save the model with `tf.keras.models.save_model`
5. Load the model using `tf.keras.models.load_model('model.h5', custom_objects={'BatchMeanSquaredError': BatchMeanSquaredError})`

The loading function raises the following error 

    ValueError: Unknown loss function: BatchMeanSquaredError

**Describe the expected behavior**
I think that the bug is caused by two little issues:

***1***
In `tensorflow_core/python/keras/losses.py` the function `get(identifier)` misses the parameter `custom_objects`. The same parameter should be added when calling `deserialize`. 
Now it is:

    def get(identifier):
      [...]
      if isinstance(identifier, six.string_types):
        identifier = str(identifier)
        return deserialize(identifier)
      if isinstance(identifier, dict):
        return deserialize(identifier)
      [...]

I think it should be:

    def get(identifier, custom_objects=None):
      [...]
      if isinstance(identifier, six.string_types):
        identifier = str(identifier)
        return deserialize(identifier, custom_objects=custom_objects)
      if isinstance(identifier, dict):
        return deserialize(identifier, custom_objects=custom_objects)
      [...]

***2***
In `tensorflow_core/python/keras/saving/saving_utils.py` the function `compile_args_from_training_config(training_config, custom_objects=None)` has the parameter `custom_objects` but the same parameter is not present when calling `losses.get`:

    if isinstance(loss_config, dict) and 'class_name' in loss_config:
        loss_config = losses.get(loss_config)

I think it should be:

    if isinstance(loss_config, dict) and 'class_name' in loss_config:
        loss_config = losses.get(loss_config, custom_objects=custom_objects)

***Overall***
The problem may be caused by the coexistence of the possibility to define `tf.keras` custom losses, or custom metrics, both using functions or classes. I think it should be allowed to use only classes to have a more general behavior.
 

**Standalone code to reproduce the issue**

    from tensorflow.keras.losses import Loss
    import tensorflow as tf
    import tensorflow.keras.backend as K
    import numpy as np
    
    
    class BatchMeanSquaredError(Loss):
    
        def __init__(self, reduction='auto', name='batch_mean_squared_error'):
            super().__init__(reduction=reduction, name=name)
    
        def call(self, y_true, y_pred):
            y_pred = tf.convert_to_tensor(y_pred)
            y_true = tf.cast(y_true, y_pred.dtype)
            L = K.mean((y_pred - y_true) ** 2, axis=0)
            return L
    
    X = np.random.random((1000, 3))
    y = np.ones(shape=(1000, 3))

    model = tf.keras.Sequential(
        [
            tf.keras.layers.Dense(3, activation='relu'),
            tf.keras.layers.Dense(3, activation='relu'),
            tf.keras.layers.Dense(3)
        ]
    )
    
    bmse = BatchMeanSquaredError()
    model.compile(loss=bmse, optimizer='sgd')

    model.fit(X, y, batch_size=10, epochs=5)

    tf.keras.models.save_model(model=model, filepath='model.h5')

    custom_objects = {'BatchMeanSquaredError': BatchMeanSquaredError}
    tf.keras.models.load_model('model.h5', custom_objects=custom_objects)

**Comment by bhack** [comment]
Can I reproduce with `Standalone code to reproduce the issue`?

**Comment by mantonios107** [comment]
@bhack Is it a question? I've posted it just to reproduce the issue

**Comment by bhack** [comment]
Yes I meant can I reproduce your issue with your code?
```
from tensorflow.keras.losses import Loss
import tensorflow as tf
import tensorflow.keras.backend as K
import numpy as np


class BatchMeanSquaredError(Loss):

    def __init__(self, reduction='auto', name='batch_mean_squared_error'):
        super().__init__(reduction=reduction, name=name)

    def call(self, y_true, y_pred):
        y_pred = tf.convert_to_tensor(y_pred)
        y_true = tf.cast(y_true, y_pred.dtype)
        L = K.mean((y_pred - y_true) ** 2, axis=0)
        return L

X = np.random.random((1000, 3))
y = np.ones(shape=(1000, 3))

model = tf.keras.Sequential(
    [
        tf.keras.layers.Dense(3, activation='relu'),
        tf.keras.layers.Dense(3, activation='relu'),
        tf.keras.layers.Dense(3)
    ]
)

bmse = BatchMeanSquaredError()
model.compile(loss=bmse, optimizer='sgd')

model.fit(X, y, batch_size=10, epochs=5)

tf.keras.models.save_model(model=model, filepath='model.h5')

custom_objects = {'BatchMeanSquaredError': BatchMeanSquaredError}
tf.keras.models.load_model('model.h5', custom_objects=custom_objects)
```

**Comment by bhack** [comment]
Cause this one is running correctly on colab.

**Comment by mantonios107** [comment]
@bhack thanks for your reply.
I wrongly executed the script in tf 2.1 while i thought I was using tf 2.3. 
The problem affects only tf 2.1 and not tf 2.2 nor tf 2.3

**Comment by bhack** [comment]
@mantonios107 Ok but I am quite sure that nobody will backport this on 2.1

**Comment by amahendrakar** [comment]
@mantonios107,
I was able to reproduce the issue with [TF v2.1](https://colab.research.google.com/gist/amahendrakar/334d94a5141bb724ec29784bcc84e19d/44210-2-1.ipynb). However, the issue seems to be fixed with [TF v2.3](https://colab.research.google.com/gist/amahendrakar/a5ae9299aba523f9a90793bbb1927025/44210.ipynb) and [TF-nightly](https://colab.research.google.com/gist/amahendrakar/918e433fb6cab78f0e4aff2e0d7ee667/44210-tf-nightly.ipynb). 

> The problem affects only tf 2.1 and not tf 2.2 nor tf 2.3

In this case, could you please update TensorFlow to the latest version i.e. v2.3. Thanks!
